quic/s2n-quic-core/src/connection/error.rs

```diff
@@ -580,3 +580,108 @@ impl From<packet_protection::Error> for ProcessingError {
         Self::DecryptError
     }
 }
+
+mod metrics {
+    use super::{transport, Error};
+    use crate::event::metrics::aggregate::{
+        info::{Str, Variant},
+        AsVariant,
+    };
+
+    macro_rules! impl_variants {
```

Comment:
I think duplicating the similar-ish code in macros like this is OK, but I wonder if we can do better mid-to-long-term. Unfortunately reflection-ish access to variants is pretty painful, but it seems like what we're expressing here is essentially a mapping from value (T) to a constant struct (Variant) which is essentially compile-time `format!("{:?}", value)`.

At the very least it seems like it would be great to eliminate the id fields.. it seems like we're not actually using those in the code? Just for checking we were exhaustive? Seems like we can do that without the id fields by just hashing the string part (which is actually more important -- that's the user visible aspect!)

if we did want an id just mapping into the index of the VARIANTS array seems sufficient, so long as it doesn't need to be stable (which seems true)?

Reply:
> it seems like what we're expressing here is essentially a mapping from value (T) to a constant struct (Variant) which is essentially compile-time format!("{:?}", value).

Yeah that's the idea

> At the very least it seems like it would be great to eliminate the id fields.. it seems like we're not actually using those in the code?

It is being used by the USDT probes. The idea was to expose a stable identifier for the [bpftrace map](https://github.com/bpftrace/bpftrace/blob/master/man/adoc/bpftrace.adoc#map-keys). But I guess we can just hash (or even just have it match the `idx`) to accomplish the same thing. Or hash is fine.

> if we did want an id just mapping into the index of the VARIANTS array seems sufficient, so long as it doesn't need to be stable (which seems true)?

Yeah I think the guarantee should be that it's only stable for the lifetime of the program. In the case of the QUIC/TLS protocol error codes, I thought it might be nice to have their actual wire values for easier debugging? But it's probably more confusing than it's worth.

Reply:
I think matching up with the index in the VARIANTS arrays makes the most sense to me.